File: Exercism/javascript/palindrome-products/palindrome-products.js
'use strict'

const product = (a, b) => a * b

const isPalindromeFactor = (pair) => pair
  .reduce(product)
  .toString()
  .split('')
  .reverse()
  .join('') === pair.reduce(product).toString()

const range = function * (from, to, step) {
  for (let i = from; i <= to; i += step) {
    yield i
  }
}
const genFactors = function * (from, to, step) {
  for (const a of range(from, to - 1, step)) {
    for (const b of range(a, to, step)) {
      yield [a, b]
    }
  }
}

const filterGen = function * (iterable, fn) {
  for (const value of iterable) {
    if (fn(value)) {
      yield value
    }
  }
}

const groupFactors = (factors) => factors
  .reduce((prod, pair, __, arr) => {
    const p = pair.reduce(product)
    if (prod.includes(p)) {
      return prod
    }
    return [...prod, [p, arr.filter((sec) => sec.reduce(product) === p)]]
  }, []).sort((a, b) => a[0] - b[0])

export class Palindromes {
  static generate ({ maxFactor, minFactor = 1 }) {
    if (minFactor > maxFactor) {
      throw new Error('min must be <= max')
    }
    const factors = groupFactors([...filterGen(genFactors(minFactor, maxFactor, 1), isPalindromeFactor)])
    const resultLength = factors.length
    const def = { value: null, factors: [] }
    return resultLength > 0
      ? { 
          smallest: {
            value: factors[0][0],
            factors: factors[0][1]
          },
          largest: {
            value: factors[resultLength-1][0],
            factors: factors[resultLength-1][1]
          }
        }
      : {
          smallest: def,
          largest: def
        }
  }
}

File: Exercism/python/palindrome-products/palindrome_products.py
from itertools import groupby


def palindromes(
        min_factor: int,
        max_factor: int,
        reverse: bool = False
        ) -> iter:
    """All palindromes and its factors"""
    return map(lambda x: (x[0], list(x[1])),
               groupby(sorted(([a, b]
                       for a in range(min_factor, max_factor + 1)
                       for b in range(a, max_factor + 1)
                       if str(a * b) == str(a * b)[::-1]),
                       key=lambda n: n[0] * n[1], reverse=reverse),
                       key=lambda n: n[0] * n[1]))


def palindrome_factors(
        min_factor: int,
        max_factor: int,
        reverse: bool = False
        ) -> iter:
    """Largest or smallest palindrome factor with assertion"""
    if min_factor > max_factor:
        raise ValueError("min_factor should be smaller then max_factor")
    try:
        return next(palindromes(min_factor, max_factor, reverse=reverse))
    except StopIteration:
        return None, []


def largest(min_factor: int, max_factor: int) -> list:
    """Largest palindrome between min- and max-factor"""
    return palindrome_factors(min_factor, max_factor, True)


def smallest(min_factor: int, max_factor: int) -> list:
    """Smallest palindrome between min- and max-factor"""
    return palindrome_factors(min_factor, max_factor)
